RecoEgamma/EgammaElectronProducers/src/LowPtGsfElectronFeatures.cc

```diff
@@ -165,83 +166,317 @@ namespace lowptgsfeleid {
 
     // Shower shape vars
-    if (ele.isNonnull()) {
-      eid_shape_full5x5_HoverE = ele->full5x5_hcalOverEcal();
-      eid_shape_full5x5_r9 = ele->full5x5_r9();
-    }
+    eid_shape_full5x5_HoverE = ele.full5x5_hcalOverEcal();
+    eid_shape_full5x5_r9 = ele.full5x5_r9();
 
     // Misc
     eid_rho = rho;
 
-    if (ele.isNonnull()) {
-      eid_brem_frac = ele->fbrem();
-      core_shFracHits = (float)ele->shFracInnerHits();
+    eid_brem_frac = ele.fbrem();
+    core_shFracHits = ele.shFracInnerHits();
+
+    // Unbiased BDT from ElectronSeed
+    gsf_bdtout1 = unbiased;
+
+    // Clusters
+    if (ele.core().isNonnull()) {
+      reco::GsfTrackRef gsf = ele.core()->gsfTrack();
+      if (gsf.isNonnull()) {
+        reco::SuperClusterRef sc = ele.core()->superCluster();
+        if (sc.isNonnull()) {
+          // Propagate electron track to ECAL surface
+          double mass2 = 0.000511 * 0.000511;
+          float p2 = pow(gsf->p(), 2);
+          float energy = sqrt(mass2 + p2);
+          XYZTLorentzVector mom = XYZTLorentzVector(gsf->px(), gsf->py(), gsf->pz(), energy);
+          XYZTLorentzVector pos = XYZTLorentzVector(gsf->vx(), gsf->vy(), gsf->vz(), 0.);
+          float field_z = 3.8;
+          RawParticle particle(mom, pos);
+          BaseParticlePropagator propagator(particle, 0., 0., field_z);
+          particle.setCharge(gsf->charge());
+          propagator.propagateToEcalEntrance(true);   // true only first half loop , false more than one loop
+          bool reach_ECAL = propagator.getSuccess();  // 0 does not reach ECAL, 1 yes barrel, 2 yes endcaps
+          GlobalPoint ecal_pos(particle.x(),          // ECAL entry point for track
+                               particle.y(),
+                               particle.z());
+
+          // Track-cluster matching for most energetic clusters
+          sc_clus1_nxtal = -999;
+          sc_clus1_dphi = -999.;
+          sc_clus2_dphi = -999.;
+          sc_clus1_deta = -999.;
+          sc_clus2_deta = -999.;
+          sc_clus1_E = -999.;
+          sc_clus2_E = -999.;
+          sc_clus1_E_ov_p = -999.;
+          sc_clus2_E_ov_p = -999.;
+          trackClusterMatching(*sc,
+                               *gsf,
+                               reach_ECAL,
+                               ecal_pos,
+                               sc_clus1_nxtal,
+                               sc_clus1_dphi,
+                               sc_clus2_dphi,
+                               sc_clus1_deta,
+                               sc_clus2_deta,
+                               sc_clus1_E,
+                               sc_clus2_E,
+                               sc_clus1_E_ov_p,
+                               sc_clus2_E_ov_p);
+          sc_clus1_nxtal = (int)sc_clus1_nxtal;
+
+        }  // sc.isNonnull()
+      }    // gsf.isNonnull()
+    }      // clusters
+
+    // Out-of-range
+    eid_rho = std::clamp(eid_rho, (float)0., (float)100.);
```

Comment:
```suggestion
    eid_rho = std::clamp(eid_rho, 0.f, 100.f);
```
same for the rest of this file/PR

Reply:
Done: https://github.com/cms-sw/cmssw/pull/31220/commits/49578e0235e80835a349bd6a3c05cdcbe3c881b5